app/src/main/java/com/sdp13epfl2021/projmag/database/UserProfileDatabase.kt

```diff
@@ -0,0 +1,142 @@
+    }
+
+    /**
+     * Return the current logged user or null if none
+     *
+     * @return current logged user or null
+     */
+    private fun getUser(): FirebaseUser? = auth.currentUser
+
+
+    /**
+     * Return the document associated to the user
+     *
+     * @return document associated to the user
+     */
+    private fun getUserDoc(): DocumentReference? =
+        getUser()?.let { user ->
+            firestore
+                .collection(UserDataFirebase.ROOT)
+                .document(user.uid)
+        }
+
+    public fun uploadProfile(
+        profile: ImmutableProfile,
+        onSuccess: () -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        val profile = hashMapOf(
+            "firstName" to profile.firstName,
+            "lastName" to profile.lastName,
+            "age" to profile.age,
+            "gender" to profile.gender.name,
+            "phoneNumber" to profile.phoneNumber,
+            "role" to profile.role.name,
+            "sciper" to profile.sciper
+        )
+        val id = getUser()?.uid
+        if (id != null) {
+            firestore.collection(ROOT).document(id)
+                .set(profile)
+                .addOnSuccessListener {
+                    onSuccess()
+                }
+                .addOnFailureListener {
+                    onFailure(it)
+                }
+        } else {
+            Log.d(TAG, "Unable to get the uid from firebase")
+        }
+    }
+
+    public fun getProfile(onSuccess: (profile: ImmutableProfile?) -> Unit) {
+        val userUid = getUser()?.uid
+        if (userUid != null) {
+            val docRef = firestore.collection(ROOT).document(userUid)
+            docRef.get()
+                .addOnSuccessListener { document ->
+                    if (document != null) {
+                        val firstName = document["firstName"] as? String
+                        val lastName = document["lastName"] as? String
+                        val ageLong = document["age"]
+
+                        val gender = when (document["gender"] as? String) {
+                            Gender.MALE.name -> Gender.MALE
+                            Gender.FEMALE.name -> Gender.FEMALE
+                            else -> Gender.OTHER
+                        }
+                        val phoneNumber = document["phoneNumber"] as? String
+                        val roleString = document["role"] as? String
+                        val sciperLong = document["sciper"]
+
+                        val role = when (roleString) {
+                            Role.TEACHER.name -> Role.TEACHER
+                            Role.STUDENT.name -> Role.STUDENT
+                            else -> Role.OTHER
+                        }
+
+
+                        var age: Int = try {
+                            Integer.valueOf(ageLong.toString())
+                        } catch (excep: NumberFormatException) {
+                            0
+                        }
```

Comment:
This kind of conversion seems a little weird and we can't know if the field was invalid after this point.
If we return null when a profile is invalid, we could use use this and do null check ?
`val age: Int? = (document["age"] as? Long)?.toInt()`